modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/internal/upgrade/v3_1_4/test/SXPBlueprintAndSXPElementUpgradeProcessTest.java

```diff
@@ -0,0 +1,315 @@
+import com.liferay.portal.kernel.test.rule.DeleteAfterTestRun;
+import com.liferay.portal.kernel.test.util.GroupTestUtil;
+import com.liferay.portal.kernel.test.util.RandomTestUtil;
+import com.liferay.portal.kernel.test.util.ServiceContextTestUtil;
+import com.liferay.portal.kernel.test.util.TestPropsValues;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.LocaleUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+import com.liferay.portal.test.rule.PermissionCheckerMethodTestRule;
+import com.liferay.portal.upgrade.registry.UpgradeStepRegistrator;
+import com.liferay.portal.upgrade.test.util.UpgradeTestUtil;
+import com.liferay.search.experiences.model.SXPBlueprint;
+import com.liferay.search.experiences.model.SXPElement;
+import com.liferay.search.experiences.service.SXPBlueprintLocalService;
+import com.liferay.search.experiences.service.SXPElementLocalService;
+
+import java.util.Collections;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+import org.junit.runner.RunWith;
+
+import org.skyscreamer.jsonassert.JSONAssert;
+import org.skyscreamer.jsonassert.JSONCompareMode;
+
+/**
+ * @author Joshua Cords
+ */
+@RunWith(Arquillian.class)
+public class SXPBlueprintAndSXPElementUpgradeProcessTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new AggregateTestRule(
+			new LiferayIntegrationTestRule(),
+			PermissionCheckerMethodTestRule.INSTANCE);
+
+	@Before
+	public void setUp() throws Exception {
+		_group = GroupTestUtil.addGroup();
+		_user = TestPropsValues.getUser();
+
+		_serviceContext = ServiceContextTestUtil.getServiceContext(
+			_group, _user.getUserId());
+	}
+
+	@Test
+	public void testSXPBlueprintUpgradeProcess() throws Exception {
+		_addAssetCategories();
+
+		SXPBlueprint sxpBlueprint = _addSXPBlueprint();
+
+		_runUpgrade();
+
+		_assertSXPBlueprint(sxpBlueprint.getSXPBlueprintId());
+	}
+
+	@Test
+	public void testSXPElementUpgradeProcess() throws Exception {
+		for (String elementExternalReferenceCode :
+				_ELEMENT_EXTERNAL_REFERENCE_CODES) {
+
+			_createOldElement(elementExternalReferenceCode);
+		}
+
+		_runUpgrade();
+
+		for (String elementExternalReferenceCode :
+				_ELEMENT_EXTERNAL_REFERENCE_CODES) {
+
+			_assertElementUpgraded(elementExternalReferenceCode);
+		}
+	}
+
+	@Rule
+	public TestName testName = new TestName();
+
+	private void _addAssetCategories() throws Exception {
+		Company company = _companyLocalService.getCompany(
+			TestPropsValues.getCompanyId());
+
+		_globalGroup = company.getGroup();
+
+		AssetVocabulary assetVocabulary = AssetTestUtil.addVocabulary(
+			_globalGroup.getGroupId());
+
+		_assetCategory1 = AssetTestUtil.addCategory(
```

Comment:
I dont like those many global variables, code gets complicated to understand. If you consider valuable to the code please move `_getExpectedInstancesJSON()` to the main test, and it will be easier to remove those global variables.